### PR TITLE
add DispatchKeySet handling to Include/ExcludeDispatchKeyGuard API

### PR DESCRIPTION
**Pull request by bhosmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42658 Include/ExcludeDispatchKeySetGuard API**
* #42619 update DispatchKey::toString()

Previously we had `IncludeDispatchKeyGuard` (resp. `Exclude`) to include/exclude single DispatchKeys. 
#42610 introduced `ExcludeDispatchKeySetGuard` to exclude an entire DispatchKeySet at once, 
but here instead we just generalize `Include/ExcludeDispatchKeyGuard` to handle DispatchKeySets 
as well as DispatchKeys.

Differential Revision: [D22971426](https://our.internmc.facebook.com/intern/diff/D22971426)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
FWIW, you probably could have prevented the call site wobbling by just reusing the old name, and additively adding set functionality to the constructor ;)

**Comment by ezyang** [comment]
There is a decent chance this is BC-breaking for external integrators because I don't think there is another API for C++ dispatch key guard, this is it

**Comment by ezyang** [comment]
cc @cpuhrsch

File: c10/core/impl/LocalDispatchKeySet.cpp [hunk 1]
```diff
@@ -47,50 +47,33 @@ void _force_tls_local_dispatch_key_set(LocalDispatchKeySet key_set) {
...
 
-IncludeDispatchKeyGuard::IncludeDispatchKeyGuard(DispatchKey x)
+IncludeDispatchKeySetGuard::IncludeDispatchKeySetGuard(DispatchKeySet include)
   : tls_(&raw_local_dispatch_key_set)
-  , id_(x)
-  // NB: prev_state_ == true on Undefined makes the guard no-op
-  , prev_state_(x == DispatchKey::Undefined ? true : tls_->included().has(x)) {
-  if (!prev_state_) {
-    tls_->set_included(tls_->included().add(x));
+  , include_(include - tls_->included()) {
```
Comment (ezyang) [inline c10/core/impl/LocalDispatchKeySet.cpp:68]: Note that this is subtly semantically different than the old version, which can be observed if there is a manual change to the dispatch key set that is not RAII guarded.

Previously:

```
// suppose K1 is included
{
  IncludeDispatchKeyGuard g(K1);
  // remove K1 from included
}
// K1 is back in included in the end!
```

After your change, K1 will still be missing in the end

Comment (bhosmer) [reply c10/core/impl/LocalDispatchKeySet.cpp:68]: Are you sure that's how it used to work? Per the big comment above, I think the old logic also ignored changes made outside the include/exclude API (i.e., didn't restore a snapshot).

Just to make sure I'm not smoking crack...
```
        old                                    new
ctor    prev_state = tls_->included has key    keyset -= tls_->included
        if !prev_state tls_->included += key   tls_->included |= keyset
dtor    if !prev_state tls_->included -= key   tls_->included -= keyset
```
If this is the right reading of the old code, I think it ignored behind-the-back changes too.

**Comment by bhosmer** [comment]
> FWIW, you probably could have prevented the call site wobbling by just reusing the old name, and additively adding set functionality to the constructor ;)

Ah yeah good idea!

**Comment by bhosmer** [comment]
@ezyang took your excellent suggestion, PR now just adds DispatchKeySet handling to the original classes. I think this puts us safely back in BC territory so removed the tag. cc @cpuhrsch

File: c10/core/impl/LocalDispatchKeySet.cpp (hunk 1 above)
Comment (ezyang) [reply c10/core/impl/LocalDispatchKeySet.cpp:68]: Oh, I guess you're right! My bad.

**Pull request closed without merging**